apis/kueue/v1beta1/clusterqueue_types.go

```diff
@@ -0,0 +1,305 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ClusterQueueSpec defines the desired state of ClusterQueue
+type ClusterQueueSpec struct {
+	// resourceGroups describes groups of resources.
+	// Each resource group defines the list of resources and a list of flavors
+	// that provide quotas for these resources.
+	// Each resource and each flavor can only form part of one resource group.
+	// resourceGroups can be up to 16.
+	// +listType=atomic
+	// +kubebuilder:validation:MaxItems=16
+	ResourceGroups []ResourceGroup `json:"resourceGroups,omitempty"`
+
+	// cohort that this ClusterQueue belongs to. CQs that belong to the
+	// same cohort can borrow unused resources from each other.
+	//
+	// A CQ can be a member of a single borrowing cohort. A workload submitted
+	// to a queue referencing this CQ can borrow quota from any CQ in the cohort.
+	// Only quota for the [resource, flavor] pairs listed in the CQ can be
+	// borrowed.
+	// If empty, this ClusterQueue cannot borrow from any other ClusterQueue and
+	// vice versa.
+	//
+	// A cohort is a name that links CQs together, but it doesn't reference any
+	// object.
+	//
+	// Validation of a cohort name is equivalent to that of object names:
+	// subdomain in DNS (RFC 1123).
+	Cohort string `json:"cohort,omitempty"`
+
+	// QueueingStrategy indicates the queueing strategy of the workloads
+	// across the queues in this ClusterQueue. This field is immutable.
+	// Current Supported Strategies:
+	//
+	// - StrictFIFO: workloads are ordered strictly by creation time.
+	// Older workloads that can't be admitted will block admitting newer
+	// workloads even if they fit available quota.
+	// - BestEffortFIFO: workloads are ordered by creation time,
+	// however older workloads that can't be admitted will not block
+	// admitting newer workloads that fit existing quota.
+	//
+	// +kubebuilder:default=BestEffortFIFO
+	// +kubebuilder:validation:Enum=StrictFIFO;BestEffortFIFO
+	QueueingStrategy QueueingStrategy `json:"queueingStrategy,omitempty"`
+
+	// namespaceSelector defines which namespaces are allowed to submit workloads to
+	// this clusterQueue. Beyond this basic support for policy, an policy agent like
+	// Gatekeeper should be used to enforce more advanced policies.
+	// Defaults to null which is a nothing selector (no namespaces eligible).
+	// If set to an empty selector `{}`, then all namespaces are eligible.
+	NamespaceSelector *metav1.LabelSelector `json:"namespaceSelector,omitempty"`
```

Comment:
I don't anticipate other kind of namespaceSelector, so I don't see a strong benefit. There is also precedence in the NetworkPolicy https://kubernetes.io/docs/concepts/services-networking/network-policies/#networkpolicy-resource

Reply:
The confusion I heard once was that this field selects the Jobs submitted to the queue. It would be better if the name coveys that this is about policy.

@liggitt any thoughts on naming here?

Reply:
how does a namespace submit workloads to the cluster queue? what happens if a workload in a namespace not matched by this selector tries to use this cluster queue?

Reply:
A namespace submits workloads to a LocalQueue. A LocalQueue references a ClusterQueue.

If the selector doesn't match, the Workload gets a Condition indicating that the Workload is not admitted because it doesn't match the namespaceSelector.

Reply:
This is a basic form of policy enforcement, ideally the cluster admin sets up something like gatekeeper, but in most cases gatekeeper is an overkill.